plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs

```diff
@@ -0,0 +1,164 @@
+    BuiltinMeaning sch _ _ -> typeSchemeToType sch
+
+{- Note [Automatic derivation of type schemes]
+We use two type classes for automatic derivation of type schemes: 'KnownMonotype' and
+'KnownPolytype'. The terminology is due to https://en.wikipedia.org/wiki/Hindley%E2%80%93Milner_type_system#The_Hindley%E2%80%93Milner_type_system
+
+'KnownMonotype' and 'KnownPolytype' are responsible for deriving monomorphic and polymorphic types,
+respectively. 'KnownMonotype' turns every argument that the Haskell denotation of a builtin
+receives into a 'TypeSchemeArrow'. We extract the arguments from the type of the Haskell denotation
+using the 'GetArgs' type family. 'KnownPolytype' turns every bound variable into a 'TypeSchemeAll'.
+We extract variables from the type of the Haskell denotation using the 'ToBinds' type family
+(in particular, see the @ToBinds (TypeScheme val args res)@ type instances). Variables are
+collected in the order that they appear in (i.e. just like in Haskell). For example, processing
+a type signature like
+
+    const
+        :: ( a ~ Opaque val (TyVarRep ('TyNameRep "a" 0))
+           , b ~ Opaque val (TyVarRep ('TyNameRep "b" 1))
+           )
+        => b -> a -> b
+
+with 'ToBinds' results in the following list of bindings:
+
+    '[ 'Some ('TyNameRep "b" 1), 'Some ('TyNameRep "a" 0) ]
+
+Higher-kinded type variables are fully supported.
+
+The implementations of the 'KnownMonotype' and 'KnownPolytype' classes are structured in an
+inference-friendly manner:
+
+1. we compute @args@ using a type family ('GetArgs') in order to dispatch on the list of
+   arguments of a built-in function in a way that doesn't force us to introduce overlapping
+   instances
+2. the @a -> res@ dependency allows us not to compute @res@ using a type family like with @args@
+3. the @args res -> a@ dependency allows us not to mention @a@ in the type of 'knownMonotype'
+
+Polymorphic built-in functions are handled via automatic specialization of all Haskell type
+variables as types representing PLC type variables, as long as each Haskell variable appears as a
+left argument to @(->)@ and is not buried somewhere inside (i.e. automatic derivation can handle
+neither @f a@, @ListRep a@, nor @f Int@. Nor is @a -> b@ allowed to the left of an @(->)@.
+Where all lower-case names are Haskell type variables). We'll call functions having such types
+"simply-polymorphic". See the docs of 'EnumerateFromTo' for details.
+
+The end result is that the user only has to specify the type of the denotation of a built-in
+function and the 'TypeScheme' of the built-in function will be derived automatically. And in the
+monomorphic and simply-polymorphic cases no types need to be specified at all.
+-}
+
+type family GetArgs a :: [GHC.Type] where
+    GetArgs (a -> b) = a ': GetArgs b
+    GetArgs _        = '[]
+
+-- | A class that allows us to derive a monotype for a builtin.
+class KnownMonotype val args res a | args res -> a, a -> res where
```

Comment:
Should this and `KnownPolytype` be in `Elaborate`, since they deal with the type-scheme-guessing machinery?

Reply:
Well, it doesn't do any elaboration, it just directly reflects a Haskell type as a `TypeScheme`. E.g. it would make sense to have this and no elaboration at all (which is how things were aligned originally). I could create a separate module for this thing too, but it seems small and straightforward enough for it not to need a separate module.